beer-garden/beer-garden
# #1171 - Refresh token support

Closes #1171

This PR switches from a single token authorization model to an access and refresh token, with refresh token rotation.  The access token has a short life (15 minutes), while the refresh token is good for slightly longer (12 hours).  There are three relevant API endpoints:

* `/api/v1/token` - This was formerly the `login` endpoint.  It has been moved here for the sake of consistency with regards to token interactions.
* `/api/v1/token/refresh` - Used to receive a new access and refresh token pair by supplying an existing, valid refresh token
* `/api/v1/token/revoke` - Used to revoke the supplied refresh token and its corresponding access token. This would generally be called when logging off to ensure that nothing can use the user's refresh token once they have declared their session to be over.

# General Notes
* There's a lot of renaming / shuffling of things due to the change from `/api/v1/login` and `LoginAPI` over to `/api/v1/token` and `TokenAPI`.  
* The original idea of just storing the token identifiers on the User model didn't work out and I ended up creating a separate `UserToken` model.  The main issue was not having a clean way to automatically expire the stored tokens.  A dedicated model let me just set an expiration time in database that mongo will handle for us.  This is more flexible for potential future changes anyway.

# Test Instructions
First, with auth enabled the UI should behave effectively as it did before these changes.  The one exception is that the UI will automatically attempt to refresh the access token once it expires (15 minutes).  If you are impatient you can hand hack [this line](https://github.com/beer-garden/beer-garden/blob/bdcfd4c02631d5d237858d76525649233b074749/src/app/beer_garden/api/http/authentication/__init__.py#L154) to shorten the access token lifespan.

Tests of the API endpoints should be possible entirely via the swagger docs.  A good sequence to test would be:
* Authenticate via the `/api/v1/token` endpoint.  You should get back an access and refresh token pair.
* Use the refresh token you received above to get a new token pair via `/api/v1/token/refresh`
* Use that new refresh token to try the `/api/v1/token/revoke` endpoint.
* Now that you've revoked the token, try using it to get a new token pair via `/api/v1/token/refresh`.  You should get a 422 response since the token is no longer valid.
